lib/multipass_ex/crypto.ex

```diff
@@ -0,0 +1,129 @@
+defmodule MultipassEx.Crypto do
+  @moduledoc false
+
+  # Module for encrypting and decrypting data in compliance with the Ruby MultiPass
+  # gem.
+  #
+  # AES_CBC128 is the encryption algorithm used.
+  #
+  # __Warning__: The ruby gem uses an insecure IV and it's mirrored here to be compatible
+  # since it is not sent with the Multipass data.
+
+  require Logger
+
+  alias MultipassEx.{Coding, CryptoError, DecodingError, EncodingError, JSONDecodingError}
+
+  @algorithm :aes_cbc128
+  @block_size 16
+  @key_byte_length 16
+  @iv String.duplicate(<<0>>, 16)
+
+  @spec encrypt(map(), binary(), binary()) ::
+          {:ok, String.t()}
+          | {:error, EncodingError.t()}
+  def encrypt(data, site_key, api_key) when is_map(data) do
+    key = generate_key(site_key, api_key)
+
+    case prepare_data_for_encryption(data) do
+      {:ok, prepared_data} ->
+        encrypted_data =
+          @algorithm
+          |> :crypto.block_encrypt(key, @iv, prepared_data)
+          |> Coding.encode()
+
+        {:ok, encrypted_data}
+
+      {:error, error_struct} ->
+        {:error, error_struct}
+    end
+  end
+
+  @spec decrypt(String.t(), binary(), binary()) ::
+          {:ok, map()} | {:error, CryptoError.t() | DecodingError.t() | JSONDecodingError.t()}
+  def decrypt(data, site_key, api_key) do
+    case Coding.decode(data) do
+      {:ok, decoded_data} ->
+        decrypt_or_error(decoded_data, site_key, api_key)
+
+      {:error, error} ->
+        {:error, error}
+    end
+  end
+
+  @spec decrypt_or_error(binary(), binary(), binary()) ::
+          {:ok, map()} | {:error, CryptoError.t() | DecodingError.t()}
+  defp decrypt_or_error(data, site_key, api_key) do
+    key = generate_key(site_key, api_key)
+
+    @algorithm
+    |> :crypto.block_decrypt(key, @iv, data)
+    |> process_decrypted_data()
+    |> case do
+      {:ok, data} -> {:ok, data}
+      {:error, error_struct} -> {:error, error_struct}
+    end
+  rescue
+    ArgumentError ->
+      {:error, %CryptoError{message: "decryption failed", reason: :argument_error}}
+  end
+
+  @spec process_decrypted_data(binary()) :: {:ok, map()} | {:error, JSONDecodingError.t()}
+  defp process_decrypted_data(data) do
+    data
+    |> unpad()
+    |> Jason.decode()
```

Comment:
good idea 👍

Reply:
I have this added but I don't think it would be a JSONDecodingError. In the case of the MultipassEx domain it would be corrupted data error, or bad data? Naming ideas?

Reply:
I just left a separate comment regarding the error API on `Multipass.decode/1`. Let's continue the conversation on that thread.